internal/premis/premis.go

```diff
@@ -0,0 +1,382 @@
+package premis
+
+import (
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/beevik/etree"
+	"github.com/google/uuid"
+)
+
+const emptyPremis = `<?xml version="1.0" encoding="UTF-8"?>
+<premis:premis xmlns:premis="http://www.loc.gov/premis/v3" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.loc.gov/premis/v3 https://www.loc.gov/standards/premis/premis.xsd" version="3.0">
+</premis:premis>
+
+`
+
+type PREMISObjectEventIdentifier struct {
+	IdType  string
+	IdValue string
+}
+
+type PREMISObject struct {
+	IdType           string
+	IdValue          string
+	OriginalName     string
+	EventIdentifiers []PREMISObjectEventIdentifier
+}
+
+type PREMISEventSummary struct {
+	Type    string
+	Detail  string
+	Outcome string
+}
+
+type PREMISEvent struct {
+	IdType       string
+	IdValue      string
+	Summary      PREMISEventSummary
+	DateTime     string
+	AgentIdType  string
+	AgentIdValue string
+}
+
+type PREMISAgent struct {
+	IdType  string
+	IdValue string
+	Name    string
+	Type    string
+}
+
+func PREMISAgentDefault() PREMISAgent {
+	return PREMISAgent{
+		Type:    "software",
+		Name:    "Enduro",
+		IdType:  "url",
+		IdValue: "https://github.com/artefactual-sdps/preprocessing-base",
+	}
+}
+
+func initializePREMISFileIfNeeded(PREMISfilepath string) error {
+	_, err := os.Stat(PREMISfilepath)
+
+	if errors.Is(err, os.ErrNotExist) {
+		err = os.WriteFile(PREMISfilepath, []byte(emptyPremis), 0o600)
+		if err != nil {
+			return err
+		}
+	} else if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func writePREMISFile(doc *etree.Document, PREMISfilepath string) error {
+	doc.Indent(2)
+
+	err := doc.WriteToFile(PREMISfilepath)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func parsePREMISDocumentAndGetRoot(PREMISfilepath string) (*etree.Document, *etree.Element, error) {
+	// Parse PREMIS document and get root PREMIS element.
+	doc := etree.NewDocument()
+
+	if err := doc.ReadFromFile(PREMISfilepath); err != nil {
+		return nil, nil, err
+	}
+
+	PREMISEl := doc.FindElement("/premis:premis")
+	if PREMISEl == nil {
+		return nil, nil, errors.New("no root premis element found in document")
+	}
+
+	return doc, PREMISEl, nil
+}
+
+func AppendPREMISObjectXML(PREMISfilepath string, object PREMISObject) error {
```

Comment:
@mcantelon I realized after pair programming yesterday that the standard Go pattern for file I/O would be to pass an `io.Writer` param to a funciton like this, and write to that stream instead of writing directly to a file.  Using `io.Writer` is more flexible then direct file I/O and makes it easier to test the output without dealing with opening and closing actual files.

Reply:
Nice! Makes sense.

Reply:
I've implemented this! Let me know if it needs tweaking.